### ISSUE TITLE
The git sidepanel does not refresh on file changes/autorefresh interval too long

### ISSUE THREAD

**Issue by telamonian** [jupyterlab/jupyterlab-git#454]
ref: https://github.com/jupyterlab/jupyterlab-git/pull/448#issuecomment-554193273

Currently, when you make changes to a file on disk in a git repo, it takes a significant lag period before said changes show up in the git sidepanel UI. One quick fix would be to shorten the autorefresh interval. Even better though would be to listen for some flavor of "file changed" signal and trigger a refresh then.

Not sure exactly which signal to listen to, but I assume there's something relevant somewhere in one of:

- `@jupyterlab/docmanager`
- `@jupyterlab/docregistry`
- `@jupyterlab/filebrowser`

**Comment by kgryte** [comment]
@telamonian Unfortunately, e.g., in `@jupyterlab/filebrowser`, the same polling logic and interval is as followed here (see <https://github.com/jupyterlab/jupyterlab/blob/78e7bf2d6848fe22f2b59200c832e60f83b6ad22/packages/filebrowser/src/model.ts#L90>).

If locking up the UI is a concern, it would be nice to have a server push-based model, rather than a pull-based model (e.g., via server sent events (SSEs)), but I sense that would entail a bit more plumbing than desired and also would seem to require that the backend of this plugin be refactored to be async.

IMO I think reducing the interval and maybe making it configurable should be the interim solution. I'll submit a PR to this end...

**Comment by telamonian** [comment]
@kgryte Before you do a PR that just reduces the interval, take another look at the snippet you linked to above. In particular the `services.contents.fileChanged.connect(...)` line

**Comment by telamonian** [comment]
Tracing it out, in the same constructor that sets the autorefresh interval for the filebrowser:

```typescript
    const refreshInterval = options.refreshInterval || DEFAULT_REFRESH_INTERVAL;

    const { services } = options.manager;
    services.contents.fileChanged.connect(this._onFileChanged, this);
    services.sessions.runningChanged.connect(this._onRunningChanged, this);

    ...

    this._poll = new Poll({
      factory: () => this.cd('.'),
      frequency: {
        interval: refreshInterval,
        backoff: true,
        max: 300 * 1000
      },
      standby: 'when-hidden'
    });
```

also connects the `_onFileChanged` method (of the filebrowser model that's being constructed) to the `fileChanged` signal of `services.contents` (which is a `Contents.IManager`). In turn, `_onFileChanged` will directly ring the `refresh` method of the model's `_poll`:

```
      void this._poll.refresh();
      this._populateSessions(sessions.running());
      this._fileChanged.emit(change);
      return;
```

So we should listen for either that same `services.contents.fileChanged` or the `fileChanged` signal that the filebrowser model has. The attached method (eg `GitExtension._onFileChanged`) would then then just call `GitExtension.refreshStatus` (which in turn calls `refresh` on its own `_poll`).

**Comment by telamonian** [comment]
@kgryte Switching over to a push-based model would also be lovely, but I for one don't know exactly how to implement that. I'm even not sure if there are any extant examples in the `jupyterlab` github org, but it would be a nice pattern for us to adopt in various places.

**Comment by kgryte** [comment]
Agreed. Listening for `fileChanged` events should work in the sense that when a user modifies/adds/remove/renames/saves a file in Lab, then we can trigger a status refresh.

However, we should still modify the polling interval and allow it to be configured IMO. If I modify a file in Finder or in my local text editor or in my IDE, TMK the `fileChanged` event is not triggered. Rather, we have to wait for the next poll. And, for the file browser extension, that is non-configurable, and IMO too slow (set to `10` seconds).

In short, I argue we should probably do as you suggest **and** reduce the polling interval.

**Comment by telamonian** [comment]
For point of reference, I just tested it out, and it appears that the default filebrowser does indeed use the default refresh value of `10s`. As well, it looks like you're right about `fileChanged`: there's no deep magic running that catches changes made outside of the Jupyter contents stack.

I'm definitely onboard for listening for `fileChanged` **and** making the polling interval configurable. I'm a little less solid on the idea of reducing the polling interval by default (if it's good enough for the default filebrowser...). I don't think it's unreasonable to expect users to do a manual refresh (or just wait a tic) if they make external changes; external changes being the only valid use case for a manual refresh button should be one of our design goals.

So I guess I'd say that you can experiment with reducing the polling interval if you want, just don't make it *too* short. Say, no less than `4s`, at least for the first stab

**Comment by kgryte** [comment]
I think the main issue with a longer polling interval is that it violates user intuition in terms of a "simple" Git interface which is as simple as save. With "save" as my mental model, I feel like, if I make an edit outside of Lab, I am going to want to save/commit it immediately, not have to wait around for Lab to poll. I sense that users, especially the non-technical types, are used to immediacy, and having to wait around is going to leave them with an impression that the UI is slow and clunky, even if we are optimizing for performance(!).

Personally, I find 3 seconds a bit slow, but maybe I am too impatient. :)

**Comment by kgryte** [comment]
And yeah, I saw that the file browser is configurable; however, that is not exposed to the end user. I think if orgs are concerned about performance, by adding a refresh interval to settings, we can allow them to adjust to taste. We can just provide a sensible default (maybe 4-5s) and tell them they can adjust accordingly.

**Comment by kgryte** [comment]
Submitted a PR. :)
